mapping_parenting_tech/analysis/prototyping/clustering_descriptions.py

```diff
@@ -57,45 +76,380 @@
 reducer = umap.UMAP(n_components=2, random_state=1)
 embedding = reducer.fit_transform(description_embeddings)
 
-# %%
-# Check the shape of the reduced embedding array
-embedding.shape
-
 # %%
 # Create another low-dim embedding for clustering
 reducer_clustering = umap.UMAP(n_components=50, random_state=1)
 embedding_clustering = reducer_clustering.fit_transform(description_embeddings)
 
 # %%
 # Clustering with hdbscan
+np.random.seed(3667)
 clusterer = hdbscan.HDBSCAN(
-    min_cluster_size=6, min_samples=1, cluster_selection_method="leaf"
+    min_cluster_size=47,
+    min_samples=1,
+    cluster_selection_method="leaf",
+    prediction_data=True,
 )
 clusterer.fit(embedding_clustering)
+logging.info(
+    f"{len(set(clusterer.labels_))} clusters with {sum([bool(x) for x in clusterer.labels_ if int(x) == -1])} unassigned apps"
+)
 
 # %%
 # Prepare dataframe for visualisation
-df = descriptions.copy()
+df = details.copy()
 df["x"] = embedding[:, 0]
 df["y"] = embedding[:, 1]
-df["cluster"] = [str(x) for x in clusterer.labels_]
+df["cluster"] = [int(x) for x in clusterer.labels_]
+df["cluster_confidence"] = [str(x) for x in clusterer.probabilities_]
+
+# %%
+cluster_col = df.columns.get_loc("cluster")
+
+radius = 3
+neigh = KNeighborsClassifier(n_neighbors=radius)
+
+current_unassigned = len(df[df.cluster == -1].cluster.to_list())
+
+cluster_list = df.cluster.to_list()
+
+unassigned_indices = []
+unassigned_indices.extend([i for i, x in enumerate(cluster_list) if x == -1])
+logging.info(f"{len(unassigned_indices)} apps not in clusters")
+
+X_list = [x for i, x in enumerate(embedding_clustering) if i not in unassigned_indices]
+y_list = [i for i in cluster_list if i != -1]
+
+neigh.fit(X_list, y_list)
+
+for x in unassigned_indices:
+    df.iat[x, cluster_col] = neigh.predict([embedding_clustering[x]])
+
+print(f"{len(df[df.cluster == -1].cluster.to_list())} unassigned apps remain.")
 
 # %%
 # Visualise using altair (NB: -1=points haven't been assigned to a cluster)
 fig = (
-    alt.Chart(df, width=500, height=500)
+    alt.Chart(df.reset_index(), width=725, height=725)
     .mark_circle(size=60)
-    .encode(x="x", y="y", tooltip=["cluster", "app_id", "summary"], color="cluster")
+    .encode(x="x", y="y", tooltip=["cluster", "appId", "summary"], color="cluster:N")
 ).interactive()
 
 fig
 
 # %%
-of_interest = ["5", "13"]
-df[df.cluster.isin(of_interest)].app_id.to_list()
+apps_of_interest = [
```

Comment:
Could you please provide a comment here on how these lists of `apps_of_interest` and `extra_apps` are used in the downstream analysis?

Reply:
Have added a comment to explain - essentially these just capture apps that were flagged as particularly important/relevant, so I created these lists to confirm that they're in clusters / which clusters they're in (but having done that, I think I deleted the code...!)